### ISSUE TITLE
Multithreading to get all influencers data

### ISSUE THREAD

**Issue by kingkimohero69** [realsirjoe/instagram-scraper#93]
Hi everyone,
   You really did a great job thank you,
I have around 100 influencers to get there followers'usernames ,there followers may range between 100k to 10 million.
So i want to make code to run all 100 influencers at the same time using different accounts and proxies (I know that every ip can use the same account for 5 times in session per once, is it right?,So i can use 25 different account with 25 different IP).
This whole process will run on the same server so i can use multithreading in python.
Will it work ? or there is any limitations or another methods to do that?

**Comment by leventdeniz** [comment]
When I tested instagram scraping with proxies and multithreading instragram still was able to detect me as a bot. So if you succeed please let me know.

**Comment by kingkimohero69** [comment]
@leventdeniz  Here is the code
`import subprocess 
import time
import os
import multiprocessing 
tasks=[]
followers_num = 100
with open('Influncers_data.csv')as f:
  influncer_data = f.read().splitlines()
with open('profile.csv')as f:
  profile = f.read().splitlines()
with open('proxy.csv')as f:
  proxies = f.read().splitlines()
# for each item in the list of items
for x,y,z in zip(proxies,influncer_data,profile):
            # construct the filename; prefix or suffix optional
            file = 'get-followers-' + y + '.py'

            # open the file to be written
            fo = open(file, 'w')
            # write the content in the file including the value being passed to each; %s indicates a string
            #fo.write('%s' % value)
            fo.write("from context import Instagram\n")
            fo.write("from time import sleep\n")
            fo.write("import csv\n")
            fo.write('print("')
            fo.write(y)
            fo.write(' process is running")\n')
            fo.write("instagram = Instagram()\n")
            fo.write("instagram.set_proxies('")
            fo.write('%s' %x)
            fo.write("')\n")
            fo.write("instagram.with_credentials('")
            fo.write('%s' % z)
            fo.write("', '4861530123')\n")
            fo.write("instagram.login(force=False,two_step_verificator=True)\n")
            fo.write("sleep(2) # Delay to mimic user\n")
            fo.write("username = '")
            fo.write('%s' % y)
            fo.write("'\n")
            fo.write("followers = []\n")
            fo.write("account = instagram.get_account(username)\n")
            fo.write("sleep(1)\n")
            fo.write("followers = instagram.get_followers(account.identifier, ")
            fo.write('%s' % followers_num)
            fo.write(", 100, delayed=True) # Get 150 followers of 'kevin', 100 a time with random delay between requests\n")
            fo.write("with open('")
            fo.write('%s' % y)
            fo.write(".csv', 'w', encoding='utf8') as f:\n")
            fo.write(" f.write('Number'+','+'ID' +','+'Username'+','+'Profile picture URL'+','+'Number of published posts'+','+'Number of followers'+','+'Number of following'+','+'Is private'+','+'Is verified'+'\\n')\n ")
            fo.write("for num,follower in enumerate(followers['accounts']):\n")
            fo.write("  #print(follower)\n")
            fo.write("  account = instagram.get_account(follower.username)\n")
            fo.write("  sleep(10)\n")
            fo.write("  f.write(str(num+1) +','+str(account.identifier)+','+follower.username+','+str(account.get_profile_picture_url())+','+str(account.media_count)+','+str(account.followed_by_count)+','+str(account.follows_count)+','+str(account.is_private) +','+str(account.is_verified)+'\\n')\n ")
            fo.write('print("')
            fo.write(y)
            fo.write(' is completed")')
            #close the file
            fo.close()
for q in influncer_data:
    tasks.append('get-followers-' + q + '.py')
# print(tasks)            
def foo(task):
    os.system('python3 ' + task)
if __name__ == '__main__':
    for task in tasks:
     p = multiprocessing.Process(target=foo, args=(task,))
     p.start()
     time.sleep(45)`

I am still facing problems, i will update with some of them 
*with default parameters page size and min,max delayed time, you will experience a rate limited error.So i used page size to be 50, min delayed time=5 seconds and max delayed time = 20 seconds 
*Another error you will face when getting every username its own data like number of followers and published posts, you will get message "try again in a few minutes",I tried to make a delay between every request but i get the same message,This step does not need to login with credentials if the profile is public, i will try it and update you with the latest progress

**Comment by leventdeniz** [comment]
Yes, sounds exactly like my experience.

**Comment by kingkimohero69** [comment]
@leventdeniz , they will detect you as a bot "rate limited error" if you are using get_account() function in a for loop without delay for at least 30 seconds, rather than that it is working scraping followers usernames only with default settings in get_followers() function

**Comment by kingkimohero69** [comment]
I was able to get 12 influencers' followers usernames at the same time after 16 hours, they ranges from 300k to 600k

**Comment by unicornofdeath** [comment]
If you use `instagram.get_followers` from my experience you get Code:429 if you ask from more than 2000 followers and delays don't help that much

**Comment by kingkimohero69** [comment]
@unicornofdeath i faced this error only in 
- when you are using instgram.get_account() in for loop after you completed scraping followers, you have to run a new script looping all followers usernames and get their data with a time.sleep(30) or more.
- if you are using the same account from multipe ip locations at the same time 
Rather than these two situations, you are good to go .
The only error iam facing right now is facebook may return a html script with message "something went wrong", i think facebook has some crashes when you look deep into older followers profile

**Comment by leventdeniz** [comment]
So you have 12 Influencers and scraped their follower usernames' with 1 influencer per thread and IP? So everytime you scrape data from instagram it's with a different IP and on a different thread, right?

But if you tried to scrape multiple influencers with the same IP you would get a error code?

You would have to need reliable proxies aswell, right. All the public proxies I tried were already known to instagram.

**Comment by kingkimohero69** [comment]
@leventdeniz yes as you said, i bought proxies from myprivateproxy.net, also i did manage to get some influencers at first with same ip, but you will get some errors like "connection reset by peer", now when digging deeply in influencers followers like 1million or more you will get error "something went wrong" code 500 or 502 or 460 .., i think facebook has api failure at some data, try it after couple of days it will work.

**Comment by leventdeniz** [comment]
Ok so with this setup you are able to scrape how many influencers data per hour?

**Comment by kingkimohero69** [comment]
200k in 6 hours with default settings

**Comment by schaumju** [comment]
@kingkimohero69 any update with this? Trying to do something similar
